types/handler.go

```diff
@@ -5,4 +5,51 @@ type Handler func(ctx Context, msg Msg) Result
 
 // AnteHandler authenticates transactions, before their internal messages are handled.
 // If newCtx.IsZero(), ctx is used instead.
-type AnteHandler func(ctx Context, tx Tx, simulate bool) (newCtx Context, result Result, abort bool)
+type AnteHandler func(ctx Context, tx Tx, simulate bool) (newCtx Context, err error)
+
+// AnteDecorator wraps the next AnteHandler to perform custom pre- and post-processing.
+type AnteDecorator interface {
+	AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error)
+}
+
+// ChainDecorator chains AnteDecorators together with each element
+// wrapping over the decorators further along chain and returns a single AnteHandler.
```

Comment:
@AdityaSripal let's update this godoc to include verbose and detailed notes on the importance of the decorator ordering, especially the fact the 1st decorator must set the gas meter otherwise there can be unexpected behavior in the BaseApp.